File: solutions/python/spiral-matrix/1/spiral_matrix.py
"""Spiral matrix generator."""


def spiral_matrix(size):
    """Return a square matrix filled in clockwise spiral order."""
    matrix = [[0] * size for _ in range(size)]
    top = 0
    bottom = size - 1
    left = 0
    right = size - 1
    value = 1

    while left <= right and top <= bottom:
        for column in range(left, right + 1):
            matrix[top][column] = value
            value += 1
        top += 1

        for row in range(top, bottom + 1):
            matrix[row][right] = value
            value += 1
        right -= 1

        for column in range(right, left - 1, -1):
            matrix[bottom][column] = value
            value += 1
        bottom -= 1

        for row in range(bottom, top - 1, -1):
            matrix[row][left] = value
            value += 1
        left += 1

    return matrix

File: solutions/scala/spiral-matrix/3/src/main/scala/SpiralMatrix.scala
object SpiralMatrix:
  def spiralMatrix(size: Int): List[List[Int]] =
    if size == 0 then List.empty
    else
      val matrix = Array.fill(size, size)(0)
      var top = 0
      var bottom = size - 1
      var left = 0
      var right = size - 1
      var number = 1

      def fillTopRow(): Unit =
        for column <- left to right do
          matrix(top)(column) = number
          number += 1
        top += 1

      def fillRightColumn(): Unit =
        for row <- top to bottom do
          matrix(row)(right) = number
          number += 1
        right -= 1

      def fillBottomRow(): Unit =
        if top <= bottom then
          for column <- right to left by -1 do
            matrix(bottom)(column) = number
            number += 1
          bottom -= 1

      def fillLeftColumn(): Unit =
        if left <= right then
          for row <- bottom to top by -1 do
            matrix(row)(left) = number
            number += 1
          left += 1

      while top <= bottom && left <= right do
        fillTopRow()
        fillRightColumn()
        fillBottomRow()
        fillLeftColumn()

      matrix.map(_.toList).toList
